agent/engine/docker_task_engine.go

```diff
@@ -238,9 +238,19 @@ func (engine *DockerTaskEngine) synchronizeState() {
 
 	tasks := engine.state.AllTasks()
 	for _, task := range tasks {
+		// Update the registry credentials of container if needed
+		// The following situations should be taken care of:
+		// 1. Agent restart:
+		//     * task is not started, task needs to wait for credentials to pull
+		//     * task already in running state, no need to wait
+		// 2. ACS send new payload message update the task
 		conts, ok := engine.state.ContainerMapByArn(task.Arn)
 		if !ok {
-			engine.startTask(task)
+			if task.ShouldWaitForExecutionCredentials() {
```

Comment:
How are you accomplishing `2. ACS send new payload message update the task` here?

Reply:
This was handled in two scenario, in both cases the credentials is automatically updated in credentials manager in agent/acs when receive the new ACS payload message:
1. All the containers of the task wasn't been pulled, where `engine.state.ContainerMapByArn(task.Arn)` will return false.
  * Agent won't do anything with this kind of task on restart, there won't be managed task for this task.
  * When received the new ACS payload message, agent check there is no managed task for this task, it will call startTask and create the new managed task. Then task manager will process this task.


2. Part of the containers in the task has been pulled/created/started, where `engine.state.ContainerMapByArn(task.Arn)` will return true.
  * On restart, the task will restore all the state of the containers and store in the task object. But won't create managed task.
  * When received the new ACS payload message,  agent check there is no managed task for this task, agent will call startTask and create managed task. Then task manager will process this task.